actionview/test/unit/cache_expiry_test.rb

```diff
@@ -0,0 +1,129 @@
+# frozen_string_literal: true
+
+require "abstract_unit"
+
+class ActionViewCacheExpiryTest < ActiveSupport::TestCase
+  def test_execution_lock_can_be_acquired_and_released
+    execution_lock = ActionView::CacheExpiry::ExecutionLock.new
+
+    # It can acquire read lock:
+    execution_lock.acquire_read_lock
+    assert_equal 1, execution_lock.read_count
+    execution_lock.release_read_lock
+    assert_equal 0, execution_lock.read_count
+
+    # It can acquire write lock:
+    execution_lock.with_write_lock do
+      assert_equal 1, execution_lock.write_count
+    end
+    assert_equal 0, execution_lock.write_count
+  end
+
+  def test_execution_lock_multiple_read_locks
+    execution_lock = ActionView::CacheExpiry::ExecutionLock.new
+
+    execution_lock.acquire_read_lock
+    assert_equal 1, execution_lock.read_count
+
+    execution_lock.acquire_read_lock
+    assert_equal 2, execution_lock.read_count
+
+    execution_lock.release_read_lock
+    assert_equal 1, execution_lock.read_count
+
+    execution_lock.release_read_lock
+    assert_equal 0, execution_lock.read_count
+  end
+
+  def test_execution_lock_read_lock_upgraded_to_write_lock
+    execution_lock = ActionView::CacheExpiry::ExecutionLock.new
+
+    execution_lock.acquire_read_lock
+    assert_equal 1, execution_lock.read_count
+
+    execution_lock.with_write_lock do
+      assert_equal 1, execution_lock.write_count
+      assert_equal 1, execution_lock.read_count
```

Comment:
Shouldn't the `read_count` always be 0 inside the `with_write_lock` block?

Reply:
No, because you can upgrade a read lock to a write lock (the reentrant part of the original design).

Reply:
Right, having the write lock is considered having both and here it's tracked explicitly in both counts.